src/tactic/simps.lean

```diff
@@ -17,110 +17,318 @@ structures, projections, simp, simplifier, generates declarations
+To do this for the projection `my_structure.awesome_projection` by adding a declaration
+`my_structure.simps.awesome_projection` that is definitionally equal to
+`my_structure.awesome_projection` but has the projection in the desired (simp-normal) form.
+
+You can initialize the projections `@[simps]` uses with `initialize_simps_projections`
+(after declaring any custom projections). This is not necessary, it has the same effect
+if you just add `@[simps]` to a declaration.
+
+If you do anything to change the default projections, make sure to call either `@[simps]` or `initialize_simps_projections` in the same file as the structure declaration. Otherwise, you might
+have a file that imports the structure, but not your custom projections.
+-/
+library_note "custom simps projection"
+
+/-- Specify simps projections, see Note [custom simps projection].
+  Set `trace.simps.verbose` to true to see the generated projections. -/
+@[user_command] meta def initialize_simps_projections_cmd
+  (_ : parse $ tk "initialize_simps_projections") : parser unit := do
+  env ← get_env,
+  ns ← many ident,
+  ns.mmap' $ λ nm, do nm ← resolve_constant nm, simps_get_raw_projections env nm
+
+/--
+  Get the projections of a structure used by simps applied to the appropriate arguments.
+  Returns a list of triples (projection expression, projection name, corresponding right-hand-side).
+
+  This function does not use `tactic.mk_app` or `tactic.mk_mapp`, because the the given arguments
+  might not uniquely specify the universe levels yet.
+
+  Example output:
+  ```
+    [(`(@prod.fst.{u v} α β), `prod.fst, `(x)), (`(@prod.snd.{u v} α β), `prod.snd, `(y))]
+  ```
+-/
+meta def simps_get_projection_exprs (e : environment) (tgt : expr)
+  (rhs : expr) : tactic $ list $ expr × name × expr := do
+  let params := get_app_args tgt, -- the parameters of the structure
+  guard ((get_app_args rhs).take params.length = params) <|> fail "unreachable code (1)",
+  let str := tgt.get_app_fn.const_name,
+  projs ← e.structure_fields_full str,
+  let rhs_args := (get_app_args rhs).drop params.length, -- the fields of the object
+  guard (rhs_args.length = projs.length) <|> fail "unreachable code (2)",
+  (raw_univs, raw_exprs) ← simps_get_raw_projections e str,
+  let univs := raw_univs.zip tgt.get_app_fn.univ_levels,
+  let proj_exprs := raw_exprs.map $
+    λ raw_expr, (raw_expr.instantiate_univ_params univs).instantiate_lambdas_or_apps params,
+  return $ proj_exprs.zip $ projs.zip rhs_args
+
+/--
+  Configuration options for the `@[simps]` attribute.
+  * `attrs` specifies the list of attributes given to the generated lemmas. Default: ``[`simp]``.
+    If ``[`simp]`` is in the list, then ``[`_refl_lemma]`` is added automatically if appropriate.
+    The attributes can be either basic attributes, or user attributes without parameters.
+    Warning: *don't* use this for user attributes whose parameter type is not `unit`. The fact that
```

Comment:
My general experience with tactic writing: nobody reads the docs until you force them. I promise you people will ignore this warning and then come complaining on Zulip that things are broken. Is there any way to make things more robust so that this case gets rejected at parse time?